### ISSUE TITLE
1) Unhandled promise rejection: Error: Session with given id not found.

### ISSUE THREAD

**Issue by esbarila** [DevExpress/testcafe#7865]
### What is your Scenario?

Following up on this conversation https://github.com/DevExpress/testcafe/issues/7849#issuecomment-1621983396

Basically getting the error put in the title.

> 1) Unhandled promise rejection:   
>    Error: Session with given id not found.

Test fails with this, even if sometimes does everything right.
Test is here, you can clone the small repo https://github.com/esbarila/AutoFi-Zygmus-challenge/blob/main/testcafe/tests/tutorialspoint.js

### What is the Current behavior?

This error fails the test.

### What is the Expected behavior?

Do everything without complaining. Or giving me something I can work with.

### What is your public website URL? (or attach your complete example)

https://www.tutorialspoint.com/html/html_iframes.htm

### What is your TestCafe test code?

https://github.com/esbarila/AutoFi-Zygmus-challenge

### Your complete configuration file

https://github.com/esbarila/AutoFi-Zygmus-challenge/blob/main/testcafe/.testcaferc.js

### Your complete test report

 ✖ Performs tasks and logs results

>    1) Unhandled promise rejection:
>       
>       Error: Session with given id not found.
>       at /home/stiva/code/AutoFi-Zygmus-challenge/node_modules/chrome-remote-interface/lib/chrome.js:94:35
>       at Chrome._handleMessage
>       (/home/stiva/code/AutoFi-Zygmus-challenge/node_modules/chrome-remote-interface/lib/chrome.js:272:17)
>       at WebSocket.<anonymous>
>       (/home/stiva/code/AutoFi-Zygmus-challenge/node_modules/chrome-remote-interface/lib/chrome.js:240:22)
>       at WebSocket.emit (node:events:513:28)
>       at WebSocket.emit (node:domain:489:12)
>       at Receiver.receiverOnMessage (/home/stiva/code/AutoFi-Zygmus-challenge/node_modules/ws/lib/websocket.js:1068:20)
>       at Receiver.emit (node:events:513:28)
>       at Receiver.emit (node:domain:489:12)
>       at Receiver.dataMessage (/home/stiva/code/AutoFi-Zygmus-challenge/node_modules/ws/lib/receiver.js:517:14)
>       at Receiver.getData (/home/stiva/code/AutoFi-Zygmus-challenge/node_modules/ws/lib/receiver.js:435:17)

### Steps to Reproduce

1. run the test to completion

### TestCafe version

testcafe": "^3.0.1

### Node.js version

18.16.0

### Command-line arguments

npx testcafe chrome testcafe/tests/tutorialspoint.js

### Browser name(s) and version(s)

Chrome 114

### Platform(s) and version(s)

Ubuntu 22.04

**Comment by miherlosev** [comment]
Hi @esbarila,

The tested website is very unstable and has a lot of errors:
![](https://github.com/DevExpress/testcafe/assets/4133518/cfb70dc3-0a09-44b2-8e46-58440fcdc010)
I ran the shared test a few times and got several different results:
![](https://github.com/DevExpress/testcafe/assets/4133518/017f9ecb-a04a-4980-bd78-e4162cbb16b4)
![](https://github.com/DevExpress/testcafe/assets/4133518/a9b7e4ae-a0bb-4aac-9420-967553b7b6ff)

Could you please extract the target HTML from this site and create a more stable example that reproduces the issue?

**Comment by tgrisley** [comment]
As my issue has been merged with this one is the consensus for my site the same? That it is unstable?

**Comment by AlexKamaev** [comment]
@tgrisley,

We think that both issues have the same cause. We'll check both usage scenarios  anyway.

**Comment by Artem-Babich** [comment]
Hi @tgrisley,

We marked #7810 as a duplicate of this issue by mistake. We are marking this issue as a duplicate of #7810 now since your example contains a lot of JS errors and before we start researching, we need to replicate the problematic behavior without them. To be sure that the cause of the issue is the same, please share an updated example that illustrates the problematic behavior without these errors.

**Comment by tgrisley** [comment]
Have bumped to this version and still sometimes encounter this error, although not as frequently as before.

<img width="999" alt="Screenshot 2023-07-18 at 13 48 35" src="https://github.com/DevExpress/testcafe/assets/126177701/4e7a2863-d62f-4f20-92b0-4aade6b68616">

For now will disable native automation.

**Comment by vasilyevi** [comment]
also still have the error on 3.1.0-rc.1

**Comment by vasilyevi** [comment]
@Artem-Babich @AlexKamaev @Aleksey28 FYI

**Comment by Artem-Babich** [comment]
Hi @tgrisley  and @vasilyevi,

I just tried both your examples with the latest TestCafe version and I cannot reproduce the issue anymore. Would you please make sure that you correctly updated to 3.1.0-rc.1 and not to 3.0.1-rc.1? Also, please make sure that you use this exact TestCafe version and not a cached one. The easiest way to check this is to add the following code to your test:
```js
import { version } from 'testcafe';

console.log(version);
```

**Comment by Artem-Babich** [comment]
Hi @tgrisley,

The issue with the selector is not related to the issue with promise rejection. Please create a separate report for it.
If I understood correctly, you confirm that the promise rejection error is not reproducible anymore.

@vasilyevi: Could you also check my [comment above](https://github.com/DevExpress/testcafe/issues/7865#issuecomment-1643838597) and clarify if you can reproduce the issue with 3.1.0-rc.1?

**Comment by tgrisley** [comment]
Hi, new issue [here](https://github.com/DevExpress/testcafe/issues/7894), and yes, now no longer getting promise rejection error it seems.

**Comment by vasilyevi** [comment]
i'll check today @Artem-Babich

**Comment by Artem-Babich** [comment]
Hi @vasilyevi and @tgrisley,

It appears the issue was caused by iframes with adverse. Therefore, it may not be consistently reproducible. In theory, this issue should be fixed in 3.1.0-rc.1. If you can reproduce it again, please update the connectionResetGuard function's code in `node_modules\testcafe\lib\native-automation\request-pipeline\safe-api.js` as follows:

```js
async function connectionResetGuard(handleRequestFn, handleErrorFn) {
    var _a;
    try {
        await handleRequestFn();
    }
    catch (err) {
        console.log(err); // <- ADDED
        if (Object.values(IGNORED_ERROR_CODES).includes((_a = err === null || err === void 0 ? void 0 : err.response) === null || _a === void 0 ? void 0 : _a.code))
            return;
        handleErrorFn(err);
        throw err;
    }
}
```

And attach the resulting console output.

**Comment by tgrisley** [comment]
Yeah i seem to get it intermittently even with 3.1.0-rc.1

<img width="1119" alt="Screenshot 2023-07-21 at 10 06 37" src="https://github.com/DevExpress/testcafe/assets/126177701/4d63f0ec-8a54-4603-b8a4-dc554a246238">

Let me try what you have suggested and get back to you.

**Comment by tgrisley** [comment]
Added the log where you mentioned and doesn't seem to be surfacing anything? 

```
async function connectionResetGuard(handleRequestFn, handleErrorFn) {
    var _a;
    try {
        await handleRequestFn();
    }
    catch (err) {
        console.log('hello');
        console.log(err); // <- ADDED
        if (Object.values(IGNORED_ERROR_CODES).includes((_a = err === null || err === void 0 ? void 0 : err.response) === null || _a === void 0 ? void 0 : _a.code))
            return;
        handleErrorFn(err);
        throw err;
    }
}

<img width="1120" alt="Screenshot 2023-07-21 at 10 27 06" src="https://github.com/DevExpress/testcafe/assets/126177701/5d9039e3-bdfe-4fcb-ac89-64af9e497152">

```

**Comment by tgrisley** [comment]
[logs.txt](https://github.com/DevExpress/testcafe/files/12122264/logs.txt)

logs attached, tests that failed if you need to search:

`should have wall present as registered user`

`should have no wall present as subscriber`

thanks

**Comment by vasilyevi** [comment]
@Artem-Babich  Okay, seems like for me it happens only in case if I add `addRequestHooks` in the beginning of the tests

**Comment by tgrisley** [comment]
Struggling to get these to fail now but will keep trying. 

Here are the logs from an all passing run as i'm struggling to get a failure to occur, still seems like errors are being surfaced even in passing tests. 

Will attempt to trigger a failure and see if the logs are different. 

[logs.txt](https://github.com/DevExpress/testcafe/files/12123622/logs.txt)

**Comment by Artem-Babich** [comment]
Hi @tgrisley,

Thank you for the logs. This is exactly the place. We will research the issue.

**Comment by tgrisley** [comment]
Great, thank-you!

Also find attached video example of [this](https://github.com/DevExpress/testcafe/issues/7865#issuecomment-1645504039)


https://github.com/DevExpress/testcafe/assets/126177701/66767e3a-4eb0-4ae8-a656-cd96824fba01

**Comment by frekel** [comment]
I have the same problem. Running testcafe on my local machine I don't get the error, but running inside a docker container in bitbucket pipelines, I get it all the time:

```    - step: &run-testcafe-tests
        name: 'Run Testcafe Tests'
        image: testcafe/testcafe:3.1.0-rc.2
        script:
          - npm ci
          - cd tests/
          - mkdir screenshots
          - /opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testRegister.js;
          - /opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testLogin.js;
          - /opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testAddToCart.js;
          - /opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testCatalogFilters.js;
          - /opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testCatalogPagination.js;
          - /opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testCatalogSearch.js;
          - /opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testContactForm.js;
          - /opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testFeedChannable.js;
          - /opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testSimpleUrlWithDomElement.js;
          - /opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testSuppliersForm.js;
        artifacts:
          - tests/screenshots/**
```

With the result:
```
/opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testSimpleUrlWithDomElement.js;
27s
+ /opt/testcafe/docker/testcafe-docker.sh 'chromium --no-sandbox --disable-dev-shm-usage --window-size="1920,1200"' --screenshots-on-fails --stop-on-first-fail testSimpleUrlWithDomElement.js;
 Running tests in:
 - Chrome 115.0.0.0 / Alpine Linux 3.18_alpha20230329
 URL Status and Element Check
 ✓ Check the page /
 ✓ Check the page /wonen-tuin
 ✓ Check the page /tuin-plantenbakken
 ✖ Check the page /sale-design-meubel-outlet
   1) Unhandled promise rejection:
      Error: Session with given id not found.
      at /usr/local/lib/node_modules/testcafe/node_modules/chrome-remote-interface/lib/chrome.js:94:35
      at Chrome._handleMessage (/usr/local/lib/node_modules/testcafe/node_modules/chrome-remote-interface/lib/chrome.js:272:17)
      at WebSocket.<anonymous> (/usr/local/lib/node_modules/testcafe/node_modules/chrome-remote-interface/lib/chrome.js:240:22)
      at WebSocket.emit (node:events:514:28)
      at WebSocket.emit (node:domain:489:12)
      at Receiver.receiverOnMessage (/usr/local/lib/node_modules/testcafe/node_modules/ws/lib/websocket.js:1068:20)
      at Receiver.emit (node:events:514:28)
      at Receiver.emit (node:domain:489:12)
      at Receiver.dataMessage (/usr/local/lib/node_modules/testcafe/node_modules/ws/lib/receiver.js:517:14)
      at Receiver.getData (/usr/local/lib/node_modules/testcafe/node_modules/ws/lib/receiver.js:435:17)
      Browser: Chrome 115.0.0.0 / Alpine Linux 3.18_alpha20230329
 16/19 failed (21s)
 ```

The script is as follows:
```
/**
 *   Checks if a given element contains as certain text for one or more urls in testSimpleUrlWithDomElement.json
 */

import { Selector } from 'testcafe';
import fs from 'fs';

const HOSTNAME = process.env.HOSTNAME_NL ? process.env.HOSTNAME_NL : '<MY WEBSITE URL>';
const SOURCE_FILE = 'src/testSimpleUrlWithDomElement.json';

fixture`URL Status and Element Check`
    .page`https://${HOSTNAME}`
    .skipJsErrors(true);

const urls = JSON.parse(fs.readFileSync(SOURCE_FILE));

const randomString = generateRandomString(); // Generate the random string once

function generateRandomString() {
    return Math.random().toString(36).substring(7);
}

for (const {host, url, selector, expectedValue } of urls) {
    test(`Check the page ${url}`, async (t) => {
        let targetHost = HOSTNAME;
        if (host && host.length > 0 && host === 'www.flinders.be') {
            targetHost =  process.env.HOSTNAME_BE;
        }

        const {origin, pathname, search} = new URL(url, `https://${targetHost}`);

        const params = new URLSearchParams(search);
        params.append('testsuite', 'testcafe');
        params.append('rand', randomString); // Use the same random string for all URLs

        const targetUrl = `${origin}${pathname}${search}${search ? '&' : '?'}${params}`;
        await t.navigateTo(targetUrl);

        const targetElement = Selector(selector);
        await t.expect(targetElement.innerText).contains(expectedValue);
    });
}

```

with the json being:
```
[
  {
    "url": "/",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/wonen-tuin",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/tuin-plantenbakken",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/sale-design-meubel-outlet",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/new",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/merken",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/hay-design",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/hay-design/accessoires",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/hay-colour-crate-mix-opberger-s?c=154727",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/hay-wieltjes-voor-colour-crate-medium",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/artemide-tolomeo-lettura-vloerlamp-halo?c=16821",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/fermob-luxembourg-tuinset-tafel-207x100-tuinbank-3-stoelen",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/kantoor",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/klantadvies",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/klantadvies/retourneren/hoe-retourneer-ik-mijn-artikelen",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  },
  {
    "url": "/wooninspiratie",
    "selector": "service-title",
    "expectedValue": "Vragen of advies?"
  }
]  
```

**Comment by frekel** [comment]
@Artem-Babich Seeing some progress!! Thanks for the hard work!  :partying_face:

**Comment by vasilyevi** [comment]
Fixed to me, but still have another blocker https://github.com/DevExpress/testcafe/issues/7898

**Comment by frekel** [comment]
Fixed for me too!  Thanks @Artem-Babich for fixing, @esbarila  for creating the issue and @vasilyevi for providing all the debug information!

:heart:

**Comment by tgrisley** [comment]
Just to let you know i am now running TestCafe 3.1.0 and still sometimes get this being thrown in the browser, causing the test to fail. 

Can seemingly get around this with quarantine mode.

Artefacts attached.


<img width="1285" alt="Screenshot 2023-08-14 at 10 08 44" src="https://github.com/DevExpress/testcafe/assets/126177701/64435953-ec27-4a8f-9d43-2cd543f142f2">

https://github.com/DevExpress/testcafe/assets/126177701/71bacdaf-0917-43d0-8882-890d3532625c

**Comment by AlexKamaev** [comment]
@tgrisley
Do you still have the `Session with given id not found` issue? It is not clear from the screenshots and video. Please share more details about the behavior you describe.

**Comment by ericmittler** [comment]
> Have bumped to 3.1.0-rc.1 and still sometimes encounter this error, although not as frequently as before.
> 
> @esbarila can you confirm that 3.1.0-rc.1 fixes it on your side, or do you encounter the same?
> 
> <img alt="Screenshot 2023-07-18 at 13 48 35" width="999" src="https://user-images.githubusercontent.com/126177701/254267758-4e7a2863-d62f-4f20-92b0-4aade6b68616.png">
> For now will disable native automation.

Thanks for this. Had the same error, upgraded and it is fixed.
